File: byul/navsys/coord/coord_list.cpp
#include "coord.h"
#include "coord_list.h"
#include <vector>
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <memory>
#include <new>

struct s_coord_list {
    std::vector<coord_t> data;
};

navsys_status_t coord_list_create_ex(coord_list_t** out_list) {
    if (!out_list) return NAVSYS_STATUS_INVALID_ARGUMENT;
    try {
        std::unique_ptr<coord_list_t> list(new coord_list_t());
        *out_list = list.release();
        return NAVSYS_STATUS_OK;
    } catch (const std::bad_alloc&) {
        return NAVSYS_STATUS_OUT_OF_MEMORY;
    } catch (...) {
        return NAVSYS_STATUS_CORRUPT_STATE;
    }
}

navsys_status_t coord_list_copy_ex(
    const coord_list_t* source, coord_list_t** out_list) {
    if (!source || !out_list) return NAVSYS_STATUS_INVALID_ARGUMENT;
    try {
        std::unique_ptr<coord_list_t> copied(new coord_list_t(*source));
        *out_list = copied.release();
        return NAVSYS_STATUS_OK;
    } catch (const std::bad_alloc&) {
        return NAVSYS_STATUS_OUT_OF_MEMORY;
    } catch (...) {
        return NAVSYS_STATUS_CORRUPT_STATE;
    }
}

size_t coord_list_size(const coord_list_t* list) {
    return list ? list->data.size() : 0;
}

navsys_status_t coord_list_fetch(
    const coord_list_t* list, size_t index, coord_t* out_coord) {
    if (!list || !out_coord) return NAVSYS_STATUS_INVALID_ARGUMENT;
    if (index >= list->data.size()) return NAVSYS_STATUS_NOT_FOUND;
    *out_coord = list->data[index];
    return NAVSYS_STATUS_OK;
}

navsys_status_t coord_list_fetch_front(
    const coord_list_t* list, coord_t* out_coord) {
    return coord_list_fetch(list, 0, out_coord);
}

navsys_status_t coord_list_fetch_back(
    const coord_list_t* list, coord_t* out_coord) {
    if (!list || !out_coord) return NAVSYS_STATUS_INVALID_ARGUMENT;
    if (list->data.empty()) return NAVSYS_STATUS_NOT_FOUND;
    *out_coord = list->data.back();
    return NAVSYS_STATUS_OK;
}

navsys_status_t coord_list_push_back_ex(
    coord_list_t* list, const coord_t* coord) {
    if (!list || !coord) return NAVSYS_STATUS_INVALID_ARGUMENT;
    const coord_t copied = *coord;
    try {
        list->data.push_back(copied);
        return NAVSYS_STATUS_OK;
    } catch (const std::bad_alloc&) {
        return NAVSYS_STATUS_OUT_OF_MEMORY;
    } catch (...) {
        return NAVSYS_STATUS_CORRUPT_STATE;
    }
}

navsys_status_t coord_list_try_pop_back(
    coord_list_t* list, coord_t* out_coord) {
    if (!list || !out_coord) return NAVSYS_STATUS_INVALID_ARGUMENT;
    if (list->data.empty()) return NAVSYS_STATUS_NOT_FOUND;
    const coord_t removed = list->data.back();
    list->data.pop_back();
    *out_coord = removed;
    return NAVSYS_STATUS_OK;
}

navsys_status_t coord_list_try_pop_front(
    coord_list_t* list, coord_t* out_coord) {
    if (!list || !out_coord) return NAVSYS_STATUS_INVALID_ARGUMENT;
    if (list->data.empty()) return NAVSYS_STATUS_NOT_FOUND;
    const coord_t removed = list->data.front();
    list->data.erase(list->data.begin());
    *out_coord = removed;
    return NAVSYS_STATUS_OK;
}

navsys_status_t coord_list_insert_ex(
    coord_list_t* list, size_t index, const coord_t* coord) {
    if (!list || !coord || index > list->data.size()) {
        return NAVSYS_STATUS_INVALID_ARGUMENT;
    }
    const coord_t copied = *coord;
    try {
        list->data.insert(
            list->data.begin() + static_cast<std::ptrdiff_t>(index),
            copied);
        return NAVSYS_STATUS_OK;
    } catch (const std::bad_alloc&) {
        return NAVSYS_STATUS_OUT_OF_MEMORY;
    } catch (...) {
        return NAVSYS_STATUS_CORRUPT_STATE;
    }
}

navsys_status_t coord_list_remove_at_ex(
    coord_list_t* list, size_t index, coord_t* out_removed) {
    if (!list || !out_removed) return NAVSYS_STATUS_INVALID_ARGUMENT;
    if (index >= list->data.size()) return NAVSYS_STATUS_NOT_FOUND;
    const coord_t removed = list->data[index];
    list->data.erase(
        list->data.begin() + static_cast<std::ptrdiff_t>(index));
    *out_removed = removed;
    return NAVSYS_STATUS_OK;
}

navsys_status_t coord_list_remove_value_ex(
    coord_list_t* list, const coord_t* coord, bool* out_removed) {
    if (!list || !coord || !out_removed) {
        return NAVSYS_STATUS_INVALID_ARGUMENT;
    }
    const coord_t copied = *coord;
    const auto it = std::find_if(
        list->data.begin(),
        list->data.end(),
        [&](const coord_t& item) {
            return item.x == copied.x && item.y == copied.y;
        });
    if (it == list->data.end()) {
        *out_removed = false;
        return NAVSYS_STATUS_OK;
    }
    list->data.erase(it);
    *out_removed = true;
    return NAVSYS_STATUS_OK;
}

navsys_status_t coord_list_find_ex(
    const coord_list_t* list,
    const coord_t* coord,
    size_t* out_index,
    bool* out_found) {
    if (!list || !coord || !out_index || !out_found) {
        return NAVSYS_STATUS_INVALID_ARGUMENT;
    }
    for (size_t index = 0; index < list->data.size(); ++index) {
        if (list->data[index].x == coord->x
            && list->data[index].y == coord->y) {
            *out_index = index;
            *out_found = true;
            return NAVSYS_STATUS_OK;
        }
    }
    *out_found = false;
    return NAVSYS_STATUS_OK;
}

coord_list_t* coord_list_create() {
    return new s_coord_list();
}

void coord_list_destroy(coord_list_t* list) {
    delete list;
}

coord_list_t* coord_list_copy(const coord_list_t* list) {
    if (!list) return nullptr;

    coord_list_t* copy = coord_list_create();

    int len = coord_list_length(list);
    for (int i =0 ; i < len; i++){
        coord_t c = list->data[i];
        copy->data.push_back(c);
    }
    return copy;
}


int coord_list_length(const coord_list_t* list) {
    return list ? static_cast<int>(list->data.size()) : 0;
}

bool coord_list_empty(const coord_list_t* list) {
    return !list || list->data.empty();
}

const coord_t* coord_list_get(const coord_list_t* list, int index) {
    if (!list || index < 0 || index >= static_cast<int>(list->data.size())) {
        return nullptr;
    }
    return &list->data[index];
}

const coord_t* coord_list_front(const coord_list_t* list) {
    if (!list || list->data.empty()) return nullptr;
    return &list->data.front();
}

const coord_t* coord_list_back(const coord_list_t* list) {
    if (!list || list->data.empty()) return nullptr;
    return &list->data.back();
}

int coord_list_push_back(coord_list_t* list, const coord_t* c) {
    if (!list || !c) return 0;
    list->data.push_back(*c);
    return 1;
}

coord_t coord_list_pop_back(coord_list_t* list) {
    coord_t result = { 0, 0 };
    if (list && !list->data.empty()) {
        result = list->data.back();
        list->data.pop_back();
    }
    return result;
}

coord_t coord_list_pop_front(coord_list_t* list) {
    coord_t result = { 0, 0 };
    if (list && !list->data.empty()) {
        result = list->data.front();
        list->data.erase(list->data.begin());
    }
    return result;
}

int coord_list_insert(coord_list_t* list, int index, const coord_t* c) {
    if (!list || !c || index < 0 ||
        index > static_cast<int>(list->data.size())) return 0;

    list->data.insert(list->data.begin() + index, *c);
    return 1;
}

void coord_list_remove_at(coord_list_t* list, int index) {
    if (!list || index < 0 ||
        index >= static_cast<int>(list->data.size())) return;

    list->data.erase(list->data.begin() + index);
}

void coord_list_remove_value(coord_list_t* list, const coord_t* c) {
    if (!list || !c) return;
    auto it = std::find_if(list->data.begin(),
        list->data.end(), [&](const coord_t& item) {
            return item.x == c->x && item.y == c->y;
        });
    if (it != list->data.end()) {
        list->data.erase(it);
    }
}

int coord_list_contains(const coord_list_t* list, const coord_t* c) {
    if (!list || !c) return 0;
    return std::any_of(list->data.begin(), list->data.end(),
        [&](const coord_t& item) {
            return item.x == c->x && item.y == c->y;
        }) ? 1 : 0;
}

int coord_list_find(const coord_list_t* list, const coord_t* c) {
    if (!list || !c) return -1;
    for (size_t i = 0; i < list->data.size(); ++i) {
        if (list->data[i].x == c->x && list->data[i].y == c->y)
            return static_cast<int>(i);
    }
    return -1;
}

coord_list_t* coord_list_sublist(
    const coord_list_t* list, int start, int end) {

    if (!list || start < 0 ||
        end > static_cast<int>(list->data.size()) || end <= start)
        return nullptr;

    coord_list_t* sub = new s_coord_list();
    for (int i = start; i < end; ++i) {
        sub->data.push_back(list->data[i]);
    }
    return sub;
}

bool coord_list_equals(const coord_list_t* a, const coord_list_t* b) {
    if (!a || !b || a->data.size() != b->data.size()) return false;
    for (size_t i = 0; i < a->data.size(); ++i) {
        if (a->data[i].x != b->data[i].x || a->data[i].y != b->data[i].y)
            return false;
    }
    return true;
}

void coord_list_clear(coord_list_t* list) {
    if (!list) return;
    list->data.clear();
}

void coord_list_reverse(coord_list_t* list) {
    if (list) std::reverse(list->data.begin(), list->data.end());
}

File: tools/byul_grid/byul_grid/wrapper/modules/coord_list.py
from ffi_core import ffi, C
import weakref

from coord import c_coord

ffi.cdef("""
// Opaque structure definition
typedef struct s_coord_list coord_list_t;

// Creation/Destruction
coord_list_t* coord_list_create(void);
void coord_list_destroy(coord_list_t* list);
coord_list_t* coord_list_copy(const coord_list_t* list);

// Information
int coord_list_length(const coord_list_t* list);
bool coord_list_empty(const coord_list_t* list);
const coord_t* coord_list_get(const coord_list_t* list, int index);
const coord_t* coord_list_front(const coord_list_t* list);
const coord_t* coord_list_back(const coord_list_t* list);

// Modification
int coord_list_push_back(coord_list_t* list, const coord_t* c);

/// @brief Removes and returns the last element of the list (NULL if none)
coord_t coord_list_pop_back(coord_list_t* list);

/// @brief Removes and returns the first element of the list (NULL if none)
coord_t coord_list_pop_front(coord_list_t* list);

int coord_list_insert(coord_list_t* list, int index, const coord_t* c);
void coord_list_remove_at(coord_list_t* list, int index);
void coord_list_remove_value(coord_list_t* list, const coord_t* c);
void coord_list_clear(coord_list_t* list);
void coord_list_reverse(coord_list_t* list);

// Search
int  coord_list_contains(const coord_list_t* list, const coord_t* c);  // returns 1 if contains
int  coord_list_find(const coord_list_t* list, const coord_t* c);      // returns index, -1 if not found

// Sublist extraction
coord_list_t* coord_list_sublist(const coord_list_t* list, int start, int end);

// Comparison
bool coord_list_equals(const coord_list_t* a, const coord_list_t* b);

""")

class c_coord_list:
    def __init__(self, raw_ptr=None, own=False):
        if raw_ptr is not None:
            self._c = raw_ptr
            self._own = own
        else:
            self._c = C.coord_list_create()
            if not self._c:
                raise MemoryError("coord_list allocation failed")
            self._own = True

        if self._own:
            self._finalizer = weakref.finalize(self, C.coord_list_destroy, self._c)
        else:
            self._finalizer = None

    def __len__(self):
        return C.coord_list_length(self._c)

    def __getitem__(self, index):
        if index < 0 or index >= len(self):
            raise IndexError("coord_list index out of range")
        ptr = C.coord_list_get(self._c, index)
        return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

    def __iter__(self):
        for i in range(len(self)):
            yield self[i]

    def __contains__(self, item):
        return isinstance(item, c_coord) and C.coord_list_contains(self._c, item.ptr()) != 0

    def index(self, item):
        if not isinstance(item, c_coord):
            raise TypeError("index() expects a c_coord object")
        return C.coord_list_find(self._c, item.ptr())

    def front(self):
        ptr = C.coord_list_front(self._c)
        return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

    def back(self):
        ptr = C.coord_list_back(self._c)
        return c_coord(raw_ptr=ptr) if ptr != ffi.NULL else None

    def append(self, coord):
        if not isinstance(coord, c_coord):
            raise TypeError("append expects a c_coord object")
        return C.coord_list_push_back(self._c, coord.ptr())

    def pop(self):
        val = C.coord_list_pop_back(self._c)
        return c_coord(raw_ptr=ffi.new("coord_t *", val))

    def pop_front(self):
        val = C.coord_list_pop_front(self._c)
        return c_coord(raw_ptr=ffi.new("coord_t *", val))

    def insert(self, index, coord):
        if not isinstance(coord, c_coord):
            raise TypeError("insert expects a c_coord object")
        return C.coord_list_insert(self._c, index, coord.ptr())

    def remove_at(self, index):
        C.coord_list_remove_at(self._c, index)

    def remove_value(self, coord):
        if not isinstance(coord, c_coord):
            raise TypeError("remove_value expects a c_coord object")
        C.coord_list_remove_value(self._c, coord.ptr())

    def clear(self):
        C.coord_list_clear(self._c)

    def reverse(self):
        C.coord_list_reverse(self._c)

    def copy(self):
        return c_coord_list(raw_ptr=C.coord_list_copy(self._c), own=True)

    def sublist(self, start, end):
        return c_coord_list(raw_ptr=C.coord_list_sublist(self._c, start, end), own=True)

    def equals(self, other):
        return isinstance(other, c_coord_list) and C.coord_list_equals(self._c, other._c)

    def empty(self):
        return C.coord_list_empty(self._c)

    def ptr(self):
        return self._c

    def __str__(self):
        return "[" + ", ".join(str(c) for c in self) + "]"

    def __repr__(self):
        return f"c_coord_list(len={len(self)})"

    def __del__(self):
        self.close()

    def close(self):
        if self._own and self._finalizer and self._finalizer.alive:
            self._finalizer()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def to_list(self):
        return [c.copy() for c in self]

    @classmethod
    def from_list(cls, lst):
        clist = cls()
        for c in lst:
            if not isinstance(c, c_coord):
                raise TypeError("from_list() expects only c_coord elements")
            clist.append(c)
        return clist
